src/aspire/source/__init__.py

```diff
@@ -317,27 +402,18 @@ def vol_forward(self, vol, start, num):
         all_idx = np.arange(start, min(start + num, self.n))
         im = vol_project(vol, self.rots[all_idx, :, :])
 
-        im = self.eval_filters(im, start, num)
-
-        im = Image(im).shift(self.offsets[all_idx, :]).asnumpy()
-
-        im *= np.broadcast_to(self.amplitudes[all_idx], (self.L, self.L, len(all_idx)))
+        im = self.model_pipeline.forward(Image(im), indices=all_idx)
 
         return im
 
-    def _to_starfile_loopdata(self):
-        df = self._metadata.copy()
-        df = df.rename(self._metadata_aliases, axis=1)
-        df = df.drop([str(col) for col in df.columns if not col.startswith('_')], axis=1)
-
-        return df
-
     def save(self, starfile_filepath, batch_size=1024, overwrite=False):
 
-        df = self._to_starfile_loopdata()
```

Comment:
I think we discussed this before, but ideally, this `save()` functionality should lie outside the `ImageSource`, since it enforces a specific format (the RELION STAR file format), it's not a generic function of the `ImageSource`.

Reply:
A top level `save_star` function added to `aspire.io.star`